bpf/bpf_lxc.c

```diff
@@ -1187,10 +1198,17 @@ ipv4_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label, __u8 *reason,
 			return ret2;
 	}
 
-	verdict = policy_can_access_ingress(ctx, src_label, SECLABEL,
-					    tuple.dport, tuple.nexthdr,
-					    is_untracked_fragment,
-					    &policy_match_type, &audited);
+    /* We skip policy enforcements when an endpoint connects to itself
+     * via service clusterIP.
+     */
+    if (tuple.daddr == IPV4_LOOPBACK) {
```

Comment:
If we can get away with (re-)using the loopback bit in the conntrack entry, that sounds like the simplest option. I'm not 100% sure what the conntrack entries look like for this loopback case though, whether the loopback bit is set in the correct entry such that it can be used in this way.

Reply:
An afternoon of code digging into the loopback's conntrack entries was fun... The new version uses the conntrack bit and seems to work.

On the plus side, I think I've identified a couple cases where we may be able to remove branches or even small sections of code in the NAT/conntrack. There's potential to fix existing complexity issues there :smiley: